rest/changes_api_test.go

```diff
@@ -556,6 +546,178 @@ func postChangesAdminChannelGrant(t *testing.T, rt *RestTester) {
+	//  Case 2: Delete a document in a multiple channels (e.g. mix-3), and validate it doesn't get included in backfill
+	rt.deleteDoc(pbs3.ID, pbs3.Rev)
+	rt.deleteDoc(mix3.ID, mix3.Rev)
+
+	// Test Scenario:
+	//   1. Document mix-5 is in channels PBS, HBO (rev 1)
+	//   2. Document is deleted (rev 2)
+	//   3. Document is recreated, only in channel PBS (rev 3)
+	mix5 := rt.putDoc("mix-5", `{"channel":["PBS","HBO"]}`)
+	rt.deleteDoc(mix5.ID, mix5.Rev)
+	_ = rt.putDoc(mix5.ID, `{"channel":["PBS"]}`)
+
+	// Test Scenario:
+	//   1. Document mix-6 is in channels PBS, HBO (rev 1)
+	//   2. Document is updated only be in channel HBO (rev 2)
+	//   3. Document is updated AGAIN to remove all channels (rev 3)
+	mix6 := rt.putDoc("mix-6", `{"channel":["PBS","HBO"]}`)
+	mix6 = rt.putDoc(mix6.ID, fmt.Sprintf(`{"_rev":%q, "channel":["HBO"]}`, mix6.Rev))
+	_ = rt.putDoc(mix6.ID, fmt.Sprintf(`{"_rev":%q}`, mix6.Rev))
+
+	// Test Scenario:
+	//   1. Delete abc-2 from channel ABC
+	//   2. Update abc-3 to remove from channel ABC
+	rt.deleteDoc(abc2.ID, abc2.Rev)
+	_ = rt.putDoc(abc3.ID, fmt.Sprintf(`{"_rev":%q}`, abc3.Rev))
+
+	// Issue simple changes request
+	changesResponse := rt.Send(requestByUser(http.MethodGet, "/db/_changes", "", "bernard"))
+	assertStatus(t, changesResponse, http.StatusOK)
+	log.Printf("Response:%+v", changesResponse.Body)
+
+	expectedResults := []string{
+		`{"seq":11,"id":"abc-1","changes":[{"rev":"1-0143105976caafbda3b90cf82948dc64"}]}`,
+		`{"seq":26,"id":"abc-2","deleted":true,"removed":["ABC"],"changes":[{"rev":"2-6055be21d970eb690f48452505ea02ed"}]}`,
+		`{"seq":27,"id":"abc-3","removed":["ABC"],"changes":[{"rev":"2-09b89154aa9a0e1620da0d86528d406a"}]}`,
+	}
+	var changes changesResults
+	assert.NoError(t, base.JSONUnmarshal(changesResponse.Body.Bytes(), &changes))
+	require.Equal(t, len(expectedResults), len(changes.Results))
+
+	for index, result := range changes.Results {
+		var expectedChange db.ChangeEntry
+		assert.NoError(t, base.JSONUnmarshal([]byte(expectedResults[index]), &expectedChange))
+		assert.Equal(t, expectedChange.Seq, result.Seq)
+		assert.Equal(t, expectedChange.ID, result.ID)
+		assert.Equal(t, expectedChange.Changes, result.Changes)
+		assert.Equal(t, expectedChange.Deleted, result.Deleted)
+		assert.Equal(t, expectedChange.Removed, result.Removed)
+	}
+
+	// Update the user doc to grant access to PBS, HBO in addition to ABC
+	response := rt.SendAdminRequest(http.MethodPut, "/db/_user/bernard", `{"admin_channels":["ABC", "PBS", "HBO"]}`)
+	assertStatus(t, response, http.StatusOK)
+
+	time.Sleep(500 * time.Millisecond)
+
+	// Issue a new changes request with since=last_seq ensure that user receives all records for channel PBS
+	changesResponse = rt.Send(requestByUser(http.MethodGet,
+		fmt.Sprintf("/db/_changes?since=%s", changes.Last_Seq), "", "bernard"))
+	assertStatus(t, changesResponse, http.StatusOK)
+	log.Printf("Response:%+v", changesResponse.Body)
```

Comment:
Is a WaitForChanges enough to avoid this fixed sleep? It seems like something that might fail when running on slow hardware.

Reply:
Makes sense, replaced fixed sleep with WaitForChanges to avoid potential failures when running on slow hardware.